### ISSUE TITLE
Manually specify dependencies for DCE

### ISSUE THREAD

**Issue by binki** [HaxeFoundation/haxe#8112]
When I call an extern method which accepts a type such as `IMap<String, String>`, DCE deletes methods on my type that are required by the type.

If an extern library’s method accepts a type, Haxe should assume that the extern method may access any of the members of that type. This should result in passing such types to extern methods causing all of their members to be marked keep. Otherwise, the compilation will succeed but will fail at runtime due to the extern referencing a member which was excluded by DCE.

For example, consider the following Haxe+JS code:

```haxe
import haxe.Constraints;

@:native('({x(v) {const i = v.keys(); while (i.hasNext()) {console.log(i.next());}}})')
extern class X {
    static function x(v:IMap<String, String>):Void;
}

class Test {
    static function main() {
        var map = [
            'a'=>'b',
            'c'=>'d',
        ];
        // Only works if this following line is uncommented:
        //map.keys();
        X.x(map);
    }
}
```

With the standard options on https://try.haxe.org/ , it fails. But if I force the member used by the extern library to exist by calling it from within my haxe code by uncommenting the line which calls `map.keys()`, it succeeds.

However, this shouldn’t be necessary. An extern library provides an interface/abstraction. If I have to manually reference members of types passed to the extern method as “used” so that they aren’t deleted by DCE, that means that I have to know the internals of the extern library, breaking encapsulation/abstraction. I shouldn’t have to look at the implementation of the extern library to successfully call it.

**Comment by ncannasse** [comment]
That's the whole point of DCE, and there is no way for the haxe system to know which extern library is using what without you somehow providing this information.

We could add a dce mode such as you suggest, but let's say that your Map contains some of your custom app class, then suddenly all this class, and all the other classes it references (its member fields, etc.) will be "dce-kept".

**Comment by Simn** [comment]
Not sure what a good strategy is here. I don't think we should keep everything that is assigned to an extern by default. However, I also think it shouldn't be the responsibility of the user (of an extern) to handle this. The responsibility should be on the author of the extern, but it's unclear how we would interface here.

**Comment by binki** [comment]
@ncannasse 
> That's the whole point of DCE, and there is no way for the haxe system to know which extern library is using what without you somehow providing this information.

I don’t think this is precisely true. The point of DCE is used to get a lot of the standard library or used packages to not be included in the compiler output when you just use a few classes from them. In my own projects, I try to avoid keeping any dead code in my own codebase—so removing my own dead code is not why I enable DCE.

> We could add a dce mode such as you suggest, but let's say that your Map contains some of your custom app class, then suddenly all this class, and all the other classes it references (its member fields, etc.) will be "dce-kept".

If your extern class accepts a type, the type system should only dce-kept the things directly referenced by that extern class. For example, if the called member accepts `IMap<String, ISomeExternInterface>`, then only the parts of `IMap<K, V>` and `ISomeExternInterface` in your local codebase would need to be implemented. This shouldn’t cause custom classes implementing that interface which you never reference/instantiate to be marked dce-kept. Just that, when a custom class is kept as a type at all (and when that reference is potentially reachable from the call to the extern class), the necessary parts to implement `ISomeExternInterface` must be kept.

Sorry, I know I do not fully understand how DCE work, so please correct me. However, what I do understand is that encapsulation and inter-library isolation is broken with current way externs interact with DCE. I also understand that enabling DCE is recommended/enabled/default even for codebases which do not themselves have much dead code—isn’t it a best practice? I will try to leave this discussion to those who understand Haxe and its design better—thanks @Simn .

**Comment by kevinresol** [comment]
How about extending the @:keep meta?

- `@:keep` = keep this type or field
- `@:keep(target.Type)` = when this type or field is kept, also keep `target.Type`
- `@:keep(target.Type.field)` = when this type or field is kept, also  keep `target.Type.field`

**Comment by Gama11** [comment]
That sounds a lot like `@:ifFeature`.

**Comment by ncannasse** [comment]
@binki the problem is the following. Let's say you have an extern that does the following:

```haxe
extern class Extern {
    function getMapElementsCount( m : Map<String,MyClass> ) : Int;
}
class MyClass {
     public var other : OtherClass;
}
```

Then if you automatically do what you describe, then you have to keep the whole Map implementation, but also the whole MyClass implementation, and the whole OtherClass implementation, and all other classes referenced by any of these classes member variables or method return types, because there is no way to tell what exactly your extern might access to.

So this would quickly lead to defeat entirely the purpose of DCE. Instead, using @:ifFeature or doing some additional declarations as part of your extern allows you to keep what the extern needs, and not more. Although this requires a bit more work, this allows for proper DCE.

BTW @Gama11 @Simn , we don't have documentation for `@:ifFeature` and other DCE metadata (aparte keep) as part of our DCE documentation here https://haxe.org/manual/cr-dce.html

Changing the title of the issue accordingly

**Comment by ncannasse** [comment]
Actually, we might want something that works the other way as `@:ifFeature` as well.

**Comment by Gama11** [comment]
"The other way"? `@:unlessFeature`?

**Comment by ncannasse** [comment]
I mean, `@:ifFeature` keeps the method if the parameter is kept. In the cases of externs we want to keep the parameter if the method is kept.

**Comment by binki** [comment]
@ncannasse 
> if you automatically do what you describe, then you have to keep the whole Map implementation, but also the whole MyClass implementation, and the whole OtherClass implementation, and all other classes referenced by any of these classes member variables or method return types, because there is no way to tell what exactly your extern might access to.

But it should result in correct output code and, for example, not include custom types which are fully unreferenced.

If people are writing externs and trying to play nicely with DCE, they will switch to using interfaces and base class names that declare the minimum of what they need. That’s why things like `IMap<>` exist—so you can program to interfaces and write reusable utilities which don’t need to know the underlying implementation. Then if you pass a `Map<K, V>` to something that uses it as an `IMap<K, V>`, `Map<K, V>` can be marked as having its implementation of `IMap<K, V>` dce-kept.

Additionally, it seems nonsensical to me for the extern declaration to have `MyClass` in it. The external library can’t know about classes that your own code defines, so how does your example make sense? I.e., `Extern` would only have a declaration like that if it looked like this:

```haxe
extern class OtherClass {
}
extern class MyClass {
    public var other : OtherClass;
}
extern class Extern {
    function getMapElementsCount( m : Map<String,MyClass> ) : Int;
}
```

Because it’s all extern, there’s nothing to even keep! ;-)

**Comment by RealyUniqueName** [comment]
Documenting `@:ifFeature`: https://github.com/HaxeFoundation/HaxeManual/pull/454

And I agree with @ncannasse we need something like `@:dependsOn("IMap.exists", "IMap.get", "ISomething.*")` for cases like this.

**Comment by Simn** [comment]
I think we need something more than that so the types are actually loaded. I have a [hack](https://github.com/HaxeFoundation/haxe/blob/development/std/eval/_std/Sys.hx#L91) in the eval standard library to deal with that for a specific case. Would be nice to do this differently.

**Comment by emugel** [comment]
Is this related?   Probably should not ask, but was wondering if that `@:dependsOn` isn't what I need.

I have an extern for a javascript file from bootstrap, to use in some React app built, webpack is used to load the npm: 
```haxe
@:jsRequire('bootstrap', 'Modal')
@:ifFeature('bootstrap.Modal.*')
extern class JsBootstrapModal {
    function new (element.js.html.Element, config:{});
    function hide():Void; 
    // etc
}
```
The original js file comes from https://github.com/twbs/bootstrap/blob/main/js/src/modal.js

So, if I call a `new JsBootstrapModal(null, null)` from somewhere, the js file **is** loaded.

If no constructor, but the `bootstrap.Modal` is used from the App, the js file **is not** loaded (sad times). Even @:keep on the extern doesn't work.
